test/apps/iodemo/io_demo.cc

```diff
@@ -246,45 +381,184 @@ class P2pDemoCommon : public UcxContext {
             _size = size;
         }
 
-        inline size_t size() const {
+        inline size_t capacity() const
+        {
+            return _capacity;
+        }
+
+        inline size_t size() const
+        {
             return _size;
         }
 
+    protected:
+        const size_t _capacity;
+
+    private:
+        void *_buffer;
+        size_t _size;
+    };
+
+    class Buffer {
     public:
-        const size_t         _capacity;
+        Buffer(size_t buffer_size, BufferMemoryPool<Buffer> &pool,
+               ucs_memory_type_t memory_type) :
+            _pool(pool)
+        {
+            switch (memory_type) {
+            case UCS_MEMORY_TYPE_HOST:
+                _impl = new HostBufferImpl(buffer_size);
+                break;
+#ifdef HAVE_CUDA
+            case UCS_MEMORY_TYPE_CUDA:
+                _impl = new CudaBufferImpl(buffer_size);
+                break;
+            case UCS_MEMORY_TYPE_CUDA_MANAGED:
+                _impl = new CudaManagedBufferImpl(buffer_size);
+                break;
+#endif
+            default:
+                LOG << "ERROR: Unsupported memory type requested: "
+                    << memory_type_to_string(memory_type);
+                abort();
+            }
+
+            _impl->init();
+        }
+
+        auto buffer()
+        {
+            return _impl->buffer();
+        }
+
+        auto capacity() const
+        {
+            return _impl->capacity();
+        }
+
+        auto size() const
+        {
+            return _impl->size();
+        }
+
+        inline void resize(size_t size)
+        {
+            _impl->resize(size);
+        }
+
+        void release()
+        {
+            _pool.put(this);
+        }
+
+
+        ~Buffer()
+        {
+            delete _impl;
+        }
 
     private:
-        void*                     _buffer;
-        size_t                    _size;
-        MemoryPool<Buffer, true>& _pool;
+        BufferMemoryPool<Buffer> &_pool;
+        BufferImpl *_impl;
+    };
+
+
+    class HostBufferImpl : public BufferImpl {
+    public:
+        HostBufferImpl(size_t size) : BufferImpl(size)
+        {
+        }
+
+        virtual void init() override
+        {
+            setBuffer(memalign(ALIGNMENT, _capacity));
+        }
+
+        ~HostBufferImpl()
+        {
+            free(buffer());
+        }
+    };
+
+#ifdef HAVE_CUDA
+    class CudaMemoryBufferImpl : public BufferImpl {
+    public:
+        CudaMemoryBufferImpl(size_t size) : BufferImpl(size)
+        {
+        }
+
+        virtual void init() override
+        {
+            void *tmp;
+            cudaError_t cerr = allocate(&tmp, _capacity);
+            if (cerr != cudaSuccess) {
+                tmp = NULL;
+            }
+            setBuffer(tmp);
+        }
+
+        virtual cudaError_t allocate(void **ptr, size_t size)
+        {
+            abort();
+        };
+
+        ~CudaMemoryBufferImpl()
+        {
+            cudaFree(buffer());
+        }
```

Comment:
lets define symmetric interface to `allocate` and free the buffer from the same level of inheritance

Reply:
Do you mean - defining pure virtual `free()` here and implementing them in `CudaBufferImpl` and in `CudaManagedBufferImpl`?

It would be the same implementation (code duplication) using `cudaFree()`.

Reply:
I mean to call alloc/free on the same level of inheritance, implementation can be common